### PR TITLE
Update Red Zone to latest release (03/29/21)

### LINKED ISSUE #12596: Add "Red Zone" to official map pool.
This is a request to add "Red Zone" mission to the custom mission tab of the map pool. It was a part of last year's Map Design Contest and winner.

http://resource.openra.net/maps/15952/

### PR DESCRIPTION
**Pull request by Quetip**
Making updates to this map (work in progress).

This PR adds landing and hovering mechanics.

Closes https://github.com/OpenRA/OpenRA/issues/12596.

### CODE REVIEW

**Comment by pchote** [comment]
OpenRA.Utility(1,1): Error: Actor type `truk` consumes conditions that are not granted: loaded
OpenRA.Utility(1,1): Error: Actor type `fix` consumes conditions that are not granted: build-incomplete, being-demolished
OpenRA.Utility(1,1): Warning: Actor type `hind` grants conditions that are not consumed: being-captured
OpenRA.Utility(1,1): Error: rules.yaml:230 refers to a trait field `PipType` that does not exist on `Passenger`.
OpenRA.Utility(1,1): Error: rules.yaml:283 refers to a trait field `PipCount` that does not exist on `AmmoPool`.
OpenRA.Utility(1,1): Error: rules.yaml:284 refers to a trait field `PipType` that does not exist on `AmmoPool`.
OpenRA.Utility(1,1): Error: rules.yaml:314 refers to a trait field `PipCount` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: rules.yaml:327 refers to a trait field `Scale` that does not exist on `RenderSprites`.
OpenRA.Utility(1,1): Error: rules.yaml:331 refers to a trait field `Scale` that does not exist on `RenderSprites`.
OpenRA.Utility(1,1): Error: rules.yaml:342 refers to a trait field `PipCount` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: rules.yaml:357 refers to a trait field `Scale` that does not exist on `RenderSprites`.
OpenRA.Utility(1,1): Error: rules.yaml:377 refers to a trait field `RequiresCondition` that does not exist on `Huntable`.
OpenRA.Utility(1,1): Error: rules.yaml:385 refers to a trait field `PipCount` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: rules.yaml:387 refers to a trait field `RequiresCondition` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: rules.yaml:392 refers to a trait field `RequiresCondition` that does not exist on `Huntable`.
OpenRA.Utility(1,1): Error: rules.yaml:416 refers to a trait field `RequiresCondition` that does not exist on `Huntable`.
OpenRA.Utility(1,1): Error: rules.yaml:422 refers to a trait field `PipCount` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: rules.yaml:441 refers to a trait field `RequiresCondition` that does not exist on `Huntable`.
OpenRA.Utility(1,1): Error: rules.yaml:447 refers to a trait field `PipCount` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: rules.yaml:460 refers to a trait field `RequiresCondition` that does not exist on `Huntable`.
OpenRA.Utility(1,1): Error: rules.yaml:466 refers to a trait field `PipCount` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: weapons.yaml:34 refers to a weapon field `BurstDelay` that does not exist.
OpenRA.Utility(1,1): Error: weapons.yaml:40 refers to a projectile field `Angle` that does not exist on `Bullet`.
OpenRA.Utility(1,1): Error: weapons.yaml:120 refers to a warhead field `ImpactSound` that does not exist on `CreateEffect`.

**Comment by Quetip** [comment]
acknowledged.

**Comment by Quetip** [comment]
Can I reopen this and make changes to or just open a new PR?

**Comment by abcdefg30** [comment]
Whatever suits you better. (You can do both.)

File: mods/ra/maps/red-zone/map.yaml [hunk 1]
```diff
@@ -0,0 +1,3352 @@
...
+		Enemies: Creeps, USSR, BadGuy
+	PlayerReference@England Main Base:
+		Name: England Main Base
+		NonCombatant: True
+		Faction: allies
+		LockColor: True
+		Color: ABB7E4
+		Allies: England Northern Base, England Air Base, England Prison Base, England Missile Silos
+		Enemies: Creeps, USSR, BadGuy
+	PlayerReference@Creeps:
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/map.yaml]: Creeps can be removed if they aren't in this mission.

File: mods/ra/maps/red-zone/map.yaml [hunk 2]
```diff
@@ -0,0 +1,3352 @@
...
+		Name: BadGuy
+		NonCombatant: True
+		Faction: soviet
+		LockColor: True
+		Color: FE1100
+		Allies: USSR
+		Enemies: Creeps, England Missile Silos, England Prison Base, England Main Base, England Northern Base, England Air Base
+
+Actors:
+	Actor1243: cycl
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/map.yaml]: This is a pain, but all fences and walls need to be assigned a correct owner for targeting purposes. (England I assume)

File: mods/ra/maps/red-zone/red-zone-AI.lua [hunk 3]
```diff
@@ -0,0 +1,336 @@
+﻿--[[
+   Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/red-zone-AI.lua]: 2021

File: mods/ra/maps/red-zone/red-zone.lua [hunk 4]
```diff
@@ -0,0 +1,608 @@
+﻿--[[
+   Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/red-zone.lua]: 2021

File: mods/ra/maps/red-zone/red-zone.lua [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+	end)
+
+	if ussr.HasNoRequiredUnits() then
+		MissionFailed()
+	end
+end
+
+EnglandAirBaseReveal = function()
+	if not EnglandAirBaseRevealed then
+		Media.DisplayMessage("Hijack a Hind Helicopter.")
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/red-zone.lua]: Considering we changed these to Blackhawks for Skirmish/Multiplayer I'd go ahead a switch this to the UH60.

File: mods/ra/maps/red-zone/red-zone.lua [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+	return player.IsNonCombatant == false and PlayerIsTeamAi(player) == false
+end
+
+
+WorldLoaded = function()
+	local teamPlayers = Player.GetPlayers(function(p)
+		return PlayerIsHumanOrBot(p)
+	end)
+
+	ussr = Player.GetPlayer("USSR")
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/red-zone.lua]: Non `local` references should be capitalized.

File: mods/ra/maps/red-zone/red-zone.lua [hunk 7]
```diff
@@ -0,0 +1,608 @@
...
+	SetupFactories()
+	SetupMainFactories()
+
+	Trigger.OnObjectiveAdded(ussr, function(p, id)
+		Trigger.AfterDelay(DateTime.Seconds(5), function()
+			Media.DisplayMessage(p.GetObjectiveDescription(id), "Primary Objective")
+		end)
+	end)
+
+	AlliedObjective = england_main.AddPrimaryObjective("Don't let soviets to destroy the Missile Silos.", england_main.Color)
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/red-zone.lua]: Use `AddObjective` instead. We are phasing out `AddPrimaryObjective` and `AddSecondaryObjective`.

File: mods/ra/maps/red-zone/rules.yaml [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+
+OILB:
+	-Targetable:
+	-Selectable:
+	Interactable:
+	Explodes:
+		Weapon: Atomic
+	-TooltipDescription@ally:
+	-TooltipDescription@other:
+
```
Comment (Smittytron) [inline mods/ra/maps/red-zone/rules.yaml:488]: Extra space.

**Review by Smittytron — COMMENTED** [review]
In general I want us to cut down on the custom rules and weapons as much as possible. If the mission doesn't need it to function I'd like to cut it out.

The map.yaml is a hefty 3300+ lines, which I guess is fine but if you see any actors that look superfluous I'd go ahead and make cuts there too.

**Comment by Quetip** [comment]
Right now I'm getting it to work. At the moment, I working to get the Heli to idle in the air with out it auto landing. Setting "IdleBehavior" trait to **None** will spawn the helicopter hovering in air and the user is unable to enter the unit. So, changes need to be made somehow. There's a few mechanics that have changed with the helicopter since the initial release of this map.

File: mods/ra/maps/red-zone/rules.yaml (hunk 8 above)
Comment (Quetip) [reply mods/ra/maps/red-zone/rules.yaml:488]: We can discuss/implement these changes. I'm working on the helicopter logic now.

File: mods/ra/maps/red-zone/red-zone.lua (hunk 4 above)
Comment (Quetip) [reply mods/ra/maps/red-zone/red-zone.lua]: Updated to 2022

**Comment by Quetip** [comment]
FWIW not sure If I can pick up where this left off BUT I made these updates. I could not get the Helicopter logic to work.

**Comment by PunkPun** [comment]
Closing due to inactivity. 

It also appears like this PR was corrupted, perhaps by the last commit?

**Pull request closed without merging**